internal/internal_worker_base.go

```diff
@@ -277,6 +281,18 @@ func (bw *baseWorker) pollTask() {
 	}
 }
 
+func isNonRetriableError(err error) bool {
+	if err == nil {
+		return false
+	}
+	switch err.(type) {
+	case *shared.BadRequestError,
+		*shared.ClientVersionNotSupportedError:
```

Comment:
ClientVersionNotSupportedError will not happen in prod. What is the BadRequestError that you want to protect?

Reply:
```
&gen.BadRequestError{
				Message: fmt.Sprintf("binary %v already marked as bad deployment", binaryChecksum),
			}
```
This is the one help me find the bug. When set badbinary for a domain, all workers with that badbinary should stop working, but without this fix it end up in dead loop polling same error.

I checked and believe all other BadRequest that poller can get from `PollForDecisionTask` call should also be protected. Because currently once BadRequest happens, it means something like tasklist or domain not correct/exist.

Reply:
Is this a bug on the auto reset? The auto reset happens when a workflow(marked with a bad binary checksum) try to make progress. But the BadRequestError returns from the frontend and workflow will never make progress. 

And if we do this, how can it be auto reset?

Reply:
Not a bug on auto reset. As discussed, when different binary (either older or newer) worker out, workflow will continue

Reply:
See if you can add a unit test case to test this logic (i.e. mock return bad request and verify that worker got shutdown).